pkg/server/share_manager.go

```diff
@@ -210,6 +266,107 @@ func (m *ShareManager) resizeVolume(devicePath, mountPath string) error {
 	return nil
 }
 
+func (m *ShareManager) getEnvAsInt(key string, defaultVal int) int {
+	env := os.Getenv(key)
+	if env == "" {
+		m.logger.Warnf("Failed to get expected environment variable, env %v wasn't set, defaulting to %v", key, defaultVal)
+		return defaultVal
+	}
+	value, err := strconv.Atoi(env)
+	if err != nil {
+		m.logger.Warnf("Failed to convert environment variable, %v, value %v, to an int, defaulting to %v", key, env, defaultVal)
+		return defaultVal
+	}
+	return value
+}
+
+func (m *ShareManager) getEnvAsBool(key string, defaultVal bool) bool {
+	env := os.Getenv(key)
+	if env == "" {
+		m.logger.Warnf("Failed to get expected environment variable, env %v wasn't set, defaulting to %v", key, defaultVal)
+		return defaultVal
+	}
+	value, err := strconv.ParseBool(env)
+	if err != nil {
+		m.logger.Warnf("Failed to convert environment variable, %v, value %v, to an int, defaulting to %v", key, env, defaultVal)
+		return defaultVal
+	}
+	return value
+}
+
+func (m *ShareManager) NewKubeClient() (*kubernetes.Clientset, error) {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	clientset, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	return clientset, nil
+}
+
+func (m *ShareManager) takeLease() error {
+	if m.leaseClient == nil {
+		return fmt.Errorf("kubernetes API client is unset")
+	}
+
+	lease, err := m.leaseClient.Leases(namespace).Get(m.context, m.volume.Name, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	m.lease = lease
+
+	now := time.Now()
+	currentHolder := *m.lease.Spec.HolderIdentity
+	if currentHolder != "" {
+		m.logger.Infof("Updating lease holderIdentity from %v to %v", currentHolder, m.leaseHolder)
+	}
+
+	*m.lease.Spec.HolderIdentity = m.leaseHolder
+	*m.lease.Spec.LeaseTransitions = *m.lease.Spec.LeaseTransitions + 1
+	m.lease.Spec.AcquireTime = &metav1.MicroTime{Time: now}
+	m.lease.Spec.RenewTime = &metav1.MicroTime{Time: now}
+
+	lease, err = m.leaseClient.Leases(namespace).Update(m.context, m.lease, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	m.lease = lease
+	m.logger.Infof("Took lease for volume %v as holder %v", m.volume.Name, m.leaseHolder)
+	return nil
+}
+
+func (m *ShareManager) renewLease() error {
+	m.lease.Spec.RenewTime = &metav1.MicroTime{Time: time.Now()}
+	lease, err := m.leaseClient.Leases(namespace).Update(m.context, m.lease, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+	m.lease = lease
+	return nil
+}
+
+func (m *ShareManager) runLeaseRenew() {
+	m.logger.Infof("Starting lease renewal for volume mounted at: %v", types.GetMountPath(m.volume.Name))
+	ticker := time.NewTicker(leaseRenewInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-m.context.Done():
+			m.logger.Info("NFS lease renewal is ending")
+			return
+		case <-ticker.C:
+			if err := m.renewLease(); err != nil {
+				m.logger.Warn("Failed to renew share-manager lease - expect to be terminated.")
+			}
```

Comment:
Do we need error handling (break and return error) if it fails to renew failed?

Reply:
IMO, we can just wait for termination by longhorn-manager as commented. We should get at least one more chance at an update given that longhorn-manager allows 7 seconds and we renew every 3.

Reply:
Agreed, and that makes it take the fast-failover code path, instead of the share-manager pod "completed" path.

Reply:
I see. It's fair. The longhorn-manager (control plane) should control the termination.